File: crates/taflow-core/src/stream/frac_diff_test.rs
use super::FracDiff;

#[test]
fn frac_diff_warms_and_resets() {
    let mut state = FracDiff::new(0.5, 1e-3).unwrap();
    let input: Vec<f64> = (1..=128).map(f64::from).collect();
    let first: Vec<f64> = input
        .iter()
        .map(|&v| state.append(v).unwrap_or(f64::NAN))
        .collect();
    assert!(first.iter().any(|v| v.is_finite()));
    assert_eq!(
        state.value().map(f64::to_bits),
        first.last().copied().map(f64::to_bits)
    );
    state.reset();
    let second: Vec<f64> = input
        .iter()
        .map(|&v| state.append(v).unwrap_or(f64::NAN))
        .collect();
    assert_eq!(
        first.iter().map(|v| v.to_bits()).collect::<Vec<_>>(),
        second.iter().map(|v| v.to_bits()).collect::<Vec<_>>()
    );
}

File: tests/frac_diff_test.py
import numpy as np

from taflow import FracDiff


def test_frac_diff_lifecycle() -> None:
    values = np.arange(1.0, 129.0)
    state = FracDiff(threshold=1e-3).extend(values)
    first = state.compute()
    assert np.isnan(first).any() and np.isfinite(first).any()
    state.reset().extend(values)
    np.testing.assert_array_equal(state.compute(), first)
    assert len(state) == len(values)

